bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java

```diff
@@ -341,7 +349,67 @@ public static Resource setResource(Resource target, Resource source) {
                 targetTimedEffects.setDuration(duration);
             }
         }
-
         return target;
     }
+
+    /**
+     * Merge on/dimming/color fields from light and grouped light resources.
+     *
+     * Subsequent sparse resources will be merged into the first one.
+     * Sparse resources will be merged into any prior full state resources.
+     * Full state resources will replace any prior sparse resources.
+     */
+    public static Collection<Resource> mergeLightResources(Collection<Resource> resources) {
+        Map<String, Resource> resourceIndex = new HashMap<>();
+        Iterator<Resource> iterator = resources.iterator();
+        while (iterator.hasNext()) {
+            Resource resource = iterator.next();
+            String id = resource.getId();
+            Resource indexedResource = resourceIndex.get(id);
+            if (indexedResource == null) {
+                indexedResource = resource;
+                resourceIndex.put(id, indexedResource);
+            }
+
+            if (!LIGHT_TYPES.contains(resource.getType()) || !resource.hasHSBComponent()
+                    || resource.equals(indexedResource)) {
+                continue;
+            }
+
+            // Full state takes precedence, invalidate previous sparse resources.
+            if (resource.hasFullState()) {
+                indexedResource.setOnState(null);
+                indexedResource.setDimming(null);
+                indexedResource.setColorXy(null);
+
+                resourceIndex.put(id, resource);
+                continue;
+            }
+
+            OnState onState = resource.getOnState();
+            if (onState != null) {
+                indexedResource.setOnState(onState);
+                resource.setOnState(null);
+            }
+            Dimming dimming = resource.getDimming();
+            if (dimming != null) {
+                indexedResource.setDimming(dimming);
+                resource.setDimming(null);
+            }
+            ColorXy colorXy = resource.getColorXy();
+            if (colorXy != null) {
+                indexedResource.setColorXy(colorXy);
+                resource.setColorXy(null);
+            }
+
+            if (!resource.hasItems()) {
+                iterator.remove();
+            }
+        }
+
+        // "Garbage collect" resources with no items left.
+        resources.removeIf(r -> LIGHT_TYPES.contains(r.getType()) && !r.hasItems());
```

Comment:
This is not what the method checks, it actually checks if the resource has any items of its corresponding resource type. The word "item" was taken from the the API documentation and was used deliberately as "logic" term rather than "Java" term.

Reply:
The term 'items' can be confused with OH items. We are referring to sub-DTO fields in the main DTO. So let's make rhat clear. You already accepted my proposal to change hasHSBField, so the naming should be consistent across those methods. If the hasXYXField includes more scope than hasNonHSBField would do, then rename it to whatever 'hasXYZField' most aptly describes it. 'hasItems' does not describe enough..

However depending on the other discussion, you may need to replace that method anyway with something like 'isComplexLightResource'..

Reply:
Adopted the term "field".

Reply:
PS probably even clearer would be..

- hasHSBJsonObject() resp. hasLightJsonObject()